internal/net/call/call.go

```diff
@@ -454,198 +480,276 @@ func (rc *reconnectingConnection) watchResolver(ctx context.Context, version *Ve
-	if rc.closed {
-		return nil, fmt.Errorf("Call on closed Connection")
+// State transition actions: all of these are called with rc.mu held.
+
+func (c *clientConnection) register() {
+	switch c.state {
+	case missing:
+		c.setState(disconnected)
+	case draining:
+		c.setState(active)
 	}
+}
 
-	if len(rc.endpoints) == 0 {
-		return nil, fmt.Errorf("%w: no endpoints available", Unreachable)
+func (c *clientConnection) unregister() {
+	switch c.state {
+	case disconnected, checking, idle:
+		c.setState(missing)
+	case active:
+		c.setState(draining)
 	}
+}
 
-	// Note that it is important to hold rc.mu when calling Pick(), and it's
-	// important that we index into rc.connections with addr while still
-	// holding rc.mu. Otherwise, a Pick() call could operate on a stale set of
-	// endpoints and return an endpoint that does not exist in rc.connections.
-	var balancer = rc.opts.Balancer
-	if opts.Balancer != nil {
-		balancer = opts.Balancer
-		balancer.Update(rc.endpoints)
+func (c *clientConnection) connected() {
+	switch c.state {
+	case disconnected:
+		c.setState(checking)
 	}
+}
 
-	// TODO(mwhittaker): Think about the other places where we can perform
-	// automatic retries. We need to be careful about non-idempotent
-	// operations.
-	var connectErr error
-	for i := 0; i < maxReconnectTries; i++ {
-		endpoint, err := balancer.Pick(opts)
-		if err != nil {
-			return nil, err
-		}
-		addr := endpoint.Address()
+func (c *clientConnection) checked() {
+	switch c.state {
+	case checking:
+		c.setState(idle)
+	}
+}
 
-		if conn, ok := rc.connections[addr]; !ok || conn.ended {
-			c, err := rc.reconnect(ctx, endpoint)
-			if err != nil {
-				connectErr = err
-				continue
-			}
-			rc.connections[addr] = c
-		}
+func (c *clientConnection) callstart() {
+	switch c.state {
+	case idle:
+		c.setState(active)
+	}
+}
 
-		c := rc.connections[addr]
-		c.lastID++
-		rpc.id = c.lastID
-		c.calls[rpc.id] = rpc
-		return c, nil
+func (c *clientConnection) lastdone() {
+	switch c.state {
+	case active:
+		c.setState(idle)
+	case draining:
+		c.setState(missing)
 	}
-	return nil, connectErr
 }
 
-// reconnect establishes (or re-establishes) the network connection to the server.
-// REQUIRES: rc.mu is held.
-func (rc *reconnectingConnection) reconnect(ctx context.Context, endpoint Endpoint) (*clientConnection, error) {
-	nc, err := endpoint.Dial(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("%w: %s", CommunicationError, err)
+func (c *clientConnection) fail(details string, err error) {
+	if !c.loggedShutdown {
+		c.loggedShutdown = true
+		logError(c.logger, details, err)
 	}
-	conn := &clientConnection{
-		logger:   rc.opts.Logger,
-		endpoint: endpoint,
-		c:        nc,
-		cbuf:     bufio.NewReader(nc),
-		mu:       &rc.mu,
-		version:  initialVersion, // Updated when we hear from server
-		calls:    map[uint64]*call{},
-		lastID:   0,
+	switch c.state {
+	case checking, idle, active:
+		c.setState(disconnected)
+	case draining:
+		c.setState(missing)
```

Comment:
Could you elaborate on why draining goes to missing instead of disconnected?

Reply:
My initial reasoning was as follows: for some reason the server just transiently dropped out of the resolver results. Since the connection is still around in draining mode, it should be fine to just mark it as non-drained and continue using it.

Though I am now a bit worried about a quick server restart. The old draining connection might be pointing at a dead server incarnation, so maybe we should not try to send new requests on that. Though even that will be just a partial fix since the resolver might have missed the quick restart and we would never get into a draining state. I think the right fix might be to include an incarnation number in the assignment and use the <endpoint,incarnation> pair as the key?

For now, I added a comment saying we are intentionally reusing the old connection since it is still around.